riscv-page-tables/src/page_table.rs

```diff
@@ -989,37 +1020,32 @@ impl<T: PagingMode> GuestStagePageTable<T> {
         F: FnMut(SupervisorPageAddr) -> bool,
     {
         let num_pages = PageSize::num_4k_pages(len);
-        vaddr
+        let end = vaddr
             .checked_add_pages(num_pages)
             .ok_or(Error::AddressOverflow)?;
         let mut inner = self.inner.lock();
-        // TODO: Support huge pages, making sure we're not invalidating pages that are partially
-        // covered by the range.
-        for va in vaddr.iter_from().take(num_pages as usize) {
-            use TableEntryType::*;
-            match inner.walk(va.into()) {
-                Leaf(pte) => {
-                    if !pte.level().is_leaf() {
-                        return Err(Error::PageSizeNotSupported(pte.level().leaf_page_size()));
-                    }
-                    if !pred(pte.page_addr()) {
-                        return Err(Error::PredicateFailed);
-                    }
-                }
-                _ => {
-                    return Err(Error::PageNotMapped);
-                }
+        let mut va = vaddr;
+        loop {
+            let pte = inner.get_mapped_leaf(va)?;
+            let page_size = pte.level().leaf_page_size();
+            if !pred(pte.page_addr()) {
```

Comment:
not necessary for this PR, but it's likely that we will want to pass the page size to the predicate function as well.

Reply:
Indeed! Given the predicate is usually going through the page tracker which maintains pages at a 4k granularity, we will need this information.